File: Tests/MapofAgentsCoreTests/SemanticEdgeResolverTests.swift
import Testing
@testable import MapofAgentsCore

@Test
func resolvesMachineFolderAndFolderThreadEdges() {
    let hostID = HostID(rawValue: "host-a")
    let machine = CanvasNode(
        id: NodeID(rawValue: "machine"),
        kind: .machine,
        title: "Machine",
        position: CanvasPoint(x: 0, y: 0),
        size: .machine,
        metadata: NodeMetadata(hostID: hostID, platform: .macOS, hostStatus: .connected)
    )
    let folder = CanvasNode(
        id: NodeID(rawValue: "folder"),
        kind: .folder,
        title: "Project",
        position: CanvasPoint(x: 120, y: 0),
        size: .folder,
        metadata: NodeMetadata(hostID: hostID, folderPath: "/Users/example/project")
    )
    let thread = CanvasNode(
        id: NodeID(rawValue: "thread"),
        kind: .codexThread,
        title: "Thread",
        position: CanvasPoint(x: 240, y: 0),
        size: .thread,
        metadata: NodeMetadata(
            hostID: hostID,
            threadRef: ThreadRef(
                hostID: hostID,
                threadID: "thread-1",
                cwd: "/Users/example/project/Sources"
            )
        )
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(thread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)
    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == folder.id })
    #expect(edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == thread.id })
    #expect(!edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == thread.id })
}

@Test
func resolvesWindowsFolderAndThreadEdges() {
    let hostID = HostID(rawValue: "windows-a")
    let machine = CanvasNode(
        id: NodeID(rawValue: "machine"),
        kind: .machine,
        title: "Windows",
        position: CanvasPoint(x: 0, y: 0),
        size: .machine,
        metadata: NodeMetadata(hostID: hostID, platform: .windows, hostStatus: .connected)
    )
    let folder = CanvasNode(
        id: NodeID(rawValue: "folder"),
        kind: .folder,
        title: "mapofagents",
        position: CanvasPoint(x: 120, y: 0),
        size: .folder,
        metadata: NodeMetadata(hostID: hostID, platform: .windows, folderPath: #"C:\Users\User\mapofagents"#)
    )
    let thread = CanvasNode(
        id: NodeID(rawValue: "thread"),
        kind: .codexThread,
        title: "Thread",
        position: CanvasPoint(x: 240, y: 0),
        size: .thread,
        metadata: NodeMetadata(
            hostID: hostID,
            platform: .windows,
            threadRef: ThreadRef(
                hostID: hostID,
                threadID: "thread-1",
                cwd: #"C:\Users\User\mapofagents\src"#
            )
        )
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(thread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)
    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == folder.id })
    #expect(edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == thread.id })
    #expect(!edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == thread.id })
}

@Test
func resolvesMachineThreadEdgeOnlyWhenNoFolderOwnsThread() {
    let hostID = HostID(rawValue: "host-a")
    let machine = CanvasNode(
        id: NodeID(rawValue: "machine"),
        kind: .machine,
        title: "Machine",
        position: CanvasPoint(x: 0, y: 0),
        size: .machine,
        metadata: NodeMetadata(hostID: hostID, platform: .macOS, hostStatus: .connected)
    )
    let folder = CanvasNode(
        id: NodeID(rawValue: "folder"),
        kind: .folder,
        title: "Project",
        position: CanvasPoint(x: 120, y: 0),
        size: .folder,
        metadata: NodeMetadata(hostID: hostID, folderPath: "/Users/example/project")
    )
    let nonProjectThread = CanvasNode(
        id: NodeID(rawValue: "thread"),
        kind: .codexThread,
        title: "Scratch Thread",
        position: CanvasPoint(x: 240, y: 0),
        size: .thread,
        metadata: NodeMetadata(
            hostID: hostID,
            threadRef: ThreadRef(
                hostID: hostID,
                threadID: "thread-1",
                cwd: "/Users/example/Downloads"
            )
        )
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(nonProjectThread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)
    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == folder.id })
    #expect(edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == nonProjectThread.id })
    #expect(!edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == nonProjectThread.id })
}

@Test
func subagentThreadsDoNotReceiveSemanticParentEdges() {
    let hostID = HostID(rawValue: "host-a")
    let machine = semanticTestMachine(hostID: hostID)
    let folder = semanticTestFolder(
        id: NodeID(rawValue: "folder"),
        hostID: hostID,
        path: "/Users/example/project"
    )
    let subagent = semanticTestThread(
        id: NodeID(rawValue: "subagent"),
        hostID: hostID,
        threadID: "subagent-1",
        cwd: "/Users/example/project/worker",
        threadKind: .subagent
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(subagent)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)

    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == folder.id })
    #expect(!edges.contains { $0.kind == .folderThread && $0.target == subagent.id })
    #expect(!edges.contains { $0.kind == .machineThread && $0.target == subagent.id })
}

@Test
func resolvesOnlyNearestFolderThreadEdgeForNestedFolders() {
    let hostID = HostID(rawValue: "host-a")
    let machine = semanticTestMachine(hostID: hostID)
    let rootFolder = semanticTestFolder(
        id: NodeID(rawValue: "folder-root"),
        hostID: hostID,
        path: "/Users/example/project"
    )
    let nestedFolder = semanticTestFolder(
        id: NodeID(rawValue: "folder-nested"),
        hostID: hostID,
        path: "/Users/example/project/apps/api"
    )
    let thread = semanticTestThread(
        id: NodeID(rawValue: "thread"),
        hostID: hostID,
        threadID: "thread-1",
        cwd: "/Users/example/project/apps/api/Sources"
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(rootFolder)
    graph.upsertNode(nestedFolder)
    graph.upsertNode(thread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)

    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == rootFolder.id })
    #expect(edges.contains { $0.kind == .machineFolder && $0.source == machine.id && $0.target == nestedFolder.id })
    #expect(edges.contains { $0.kind == .folderThread && $0.source == nestedFolder.id && $0.target == thread.id })
    #expect(!edges.contains { $0.kind == .folderThread && $0.source == rootFolder.id && $0.target == thread.id })
    #expect(!edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == thread.id })
}

@Test
func folderThreadEdgesDoNotMatchSiblingPathPrefixes() {
    let hostID = HostID(rawValue: "host-a")
    let machine = semanticTestMachine(hostID: hostID)
    let folder = semanticTestFolder(
        id: NodeID(rawValue: "folder"),
        hostID: hostID,
        path: "/Users/example/project/app"
    )
    let siblingThread = semanticTestThread(
        id: NodeID(rawValue: "thread"),
        hostID: hostID,
        threadID: "thread-1",
        cwd: "/Users/example/project/application"
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(siblingThread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)

    #expect(!edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == siblingThread.id })
    #expect(edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == siblingThread.id })
}

@Test
func resolvesUncFolderThreadEdgesCaseInsensitivelyWithoutSiblingPrefixMatches() {
    let hostID = HostID(rawValue: "windows-a")
    let machine = semanticTestMachine(hostID: hostID, platform: .windows)
    let folder = semanticTestFolder(
        id: NodeID(rawValue: "folder"),
        hostID: hostID,
        platform: .windows,
        path: #"\\Server\Share\Project"#
    )
    let matchingThread = semanticTestThread(
        id: NodeID(rawValue: "matching-thread"),
        hostID: hostID,
        platform: .windows,
        threadID: "matching-thread",
        cwd: #"\\server\share\project\src"#
    )
    let siblingThread = semanticTestThread(
        id: NodeID(rawValue: "sibling-thread"),
        hostID: hostID,
        platform: .windows,
        threadID: "sibling-thread",
        cwd: #"\\server\share\project-old\src"#
    )

    var graph = AgentGraph()
    graph.upsertNode(machine)
    graph.upsertNode(folder)
    graph.upsertNode(matchingThread)
    graph.upsertNode(siblingThread)

    let edges = DefaultSemanticEdgeResolver().resolveEdges(in: graph)

    #expect(edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == matchingThread.id })
    #expect(!edges.contains { $0.kind == .folderThread && $0.source == folder.id && $0.target == siblingThread.id })
    #expect(edges.contains { $0.kind == .machineThread && $0.source == machine.id && $0.target == siblingThread.id })
}

private func semanticTestMachine(
    hostID: HostID,
    platform: HostPlatform = .macOS
) -> CanvasNode {
    CanvasNode(
        id: NodeID(rawValue: "machine"),
        kind: .machine,
        title: "Machine",
        position: CanvasPoint(x: 0, y: 0),
        size: .machine,
        metadata: NodeMetadata(hostID: hostID, platform: platform, hostStatus: .connected)
    )
}

private func semanticTestFolder(
    id: NodeID,
    hostID: HostID,
    platform: HostPlatform = .macOS,
    path: String
) -> CanvasNode {
    CanvasNode(
        id: id,
        kind: .folder,
        title: "Project",
        position: CanvasPoint(x: 120, y: 0),
        size: .folder,
        metadata: NodeMetadata(hostID: hostID, platform: platform, folderPath: path)
    )
}

private func semanticTestThread(
    id: NodeID,
    hostID: HostID,
    platform: HostPlatform = .macOS,
    threadID: String,
    cwd: String,
    threadKind: CodexThreadNodeKind? = nil
) -> CanvasNode {
    CanvasNode(
        id: id,
        kind: .codexThread,
        title: "Thread",
        position: CanvasPoint(x: 240, y: 0),
        size: .thread,
        metadata: NodeMetadata(
            hostID: hostID,
            platform: platform,
            threadRef: ThreadRef(hostID: hostID, threadID: threadID, cwd: cwd),
            threadKind: threadKind
        )
    )
}

File: windows/tests/MapofAgents.Windows.Tests/SemanticEdgeResolverTests.cs
using MapofAgents.Core;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace MapofAgents.Windows.Tests;

[TestClass]
public sealed class SemanticEdgeResolverTests
{
    [TestMethod]
    public void ResolvesMachineFolderAndMostSpecificFolderThreadEdges()
    {
        var graph = GraphWith(
            Machine("machine", "host-a"),
            Folder("root", "host-a", @"C:\Users\example"),
            Folder("repo", "host-a", @"C:\Users\example\Repo"),
            Thread("thread", "host-a", @"c:\users\example\repo\src"));

        var edges = SemanticEdgeResolver.ResolveEdges(graph);

        AssertEdge(edges, "machine", "root", EdgeKinds.MachineFolder);
        AssertEdge(edges, "machine", "repo", EdgeKinds.MachineFolder);
        AssertEdge(edges, "repo", "thread", EdgeKinds.FolderThread);
        AssertNoEdge(edges, "machine", "thread", EdgeKinds.MachineThread);
    }

    [TestMethod]
    public void ResolvesMachineThreadWhenNoFolderMatches()
    {
        var graph = GraphWith(
            Machine("machine", "host-a"),
            Folder("repo", "host-a", @"C:\Users\example\Repo"),
            Thread("thread", "host-a", @"D:\scratch"));

        var edges = SemanticEdgeResolver.ResolveEdges(graph);

        AssertEdge(edges, "machine", "repo", EdgeKinds.MachineFolder);
        AssertEdge(edges, "machine", "thread", EdgeKinds.MachineThread);
        AssertNoEdge(edges, "repo", "thread", EdgeKinds.FolderThread);
    }

    [TestMethod]
    public void SkipsSubagentThreadsLikeMacSemanticResolver()
    {
        var graph = GraphWith(
            Machine("machine", "host-a"),
            Folder("repo", "host-a", "/Users/example/repo"),
            Thread("agent", "host-a", "/Users/example/repo", ThreadKinds.Subagent));

        var edges = SemanticEdgeResolver.ResolveEdges(graph);

        AssertEdge(edges, "machine", "repo", EdgeKinds.MachineFolder);
        AssertNoEdge(edges, "repo", "agent", EdgeKinds.FolderThread);
        AssertNoEdge(edges, "machine", "agent", EdgeKinds.MachineThread);
    }

    [TestMethod]
    public void DeduplicatesEdgesAlreadyStoredInWindowsManualEdges()
    {
        var graph = GraphWith(
            Machine("machine", "host-a"),
            Folder("repo", "host-a", @"C:\Users\example\Repo"),
            Thread("thread", "host-a", @"C:\Users\example\Repo"));
        graph.ManualEdges["existing-folder-thread"] = new CanvasEdge
        {
            Id = "existing-folder-thread",
            Source = "repo",
            Target = "thread",
            Kind = EdgeKinds.FolderThread,
            IsManual = false
        };

        var semanticEdges = SemanticEdgeResolver.ResolveEdges(graph);
        var allEdges = SemanticEdgeResolver.AllEdges(graph);

        AssertNoEdge(semanticEdges, "repo", "thread", EdgeKinds.FolderThread);
        Assert.AreEqual(1, allEdges.Count(edge =>
            edge.Source == "repo" &&
            edge.Target == "thread" &&
            edge.Kind == EdgeKinds.FolderThread));
    }

    private static AgentGraph GraphWith(params CanvasNode[] nodes)
    {
        var graph = new AgentGraph();
        foreach (var node in nodes)
        {
            graph.Nodes[node.Id] = node;
        }

        return graph;
    }

    private static CanvasNode Machine(string id, string hostId)
    {
        return new CanvasNode
        {
            Id = id,
            Kind = NodeKinds.Machine,
            Title = id,
            Metadata = new NodeMetadata
            {
                HostID = hostId,
                Platform = HostPlatforms.Windows
            }
        };
    }

    private static CanvasNode Folder(string id, string hostId, string path)
    {
        return new CanvasNode
        {
            Id = id,
            Kind = NodeKinds.Folder,
            Title = id,
            Metadata = new NodeMetadata
            {
                HostID = hostId,
                Platform = HostPlatforms.Windows,
                FolderPath = path
            }
        };
    }

    private static CanvasNode Thread(
        string id,
        string hostId,
        string cwd,
        string threadKind = ThreadKinds.Thread)
    {
        return new CanvasNode
        {
            Id = id,
            Kind = NodeKinds.CodexThread,
            Title = id,
            Metadata = new NodeMetadata
            {
                HostID = hostId,
                Platform = HostPlatforms.Windows,
                ThreadKind = threadKind,
                ThreadRef = new ThreadRef
                {
                    HostID = hostId,
                    ThreadID = id,
                    Cwd = cwd,
                    Name = id
                }
            }
        };
    }

    private static void AssertEdge(
        IReadOnlyList<CanvasEdge> edges,
        string source,
        string target,
        string kind)
    {
        Assert.IsTrue(
            edges.Any(edge => edge.Source == source && edge.Target == target && edge.Kind == kind),
            $"Expected {kind} edge from {source} to {target}.");
    }

    private static void AssertNoEdge(
        IReadOnlyList<CanvasEdge> edges,
        string source,
        string target,
        string kind)
    {
        Assert.IsFalse(
            edges.Any(edge => edge.Source == source && edge.Target == target && edge.Kind == kind),
            $"Did not expect {kind} edge from {source} to {target}.");
    }
}
